stackrox/stackrox
# ROX-14299: roxctl login command

## Description

This PR adds the login command for roxctl. The login command for roxctl does the following:
- Based on the given flag values, open a browser and let the user go through the authorization flow of an auth provider of their choice.
- Obtain the token, refresh token, expiration information from Central after the authorization flow has successfully completed.

To achieve this, a new client state has been introduced explicitly for the roxctl authorize mode, similar to the test login mode we have.

Based on the auth provider type, the callback from the auth provider will be either handled within the registry handler, or the auth saga within the UI. Both parts will handle the extractions of the token and other parameters (expiration, refresh token) and then redirect back to the server opened by roxctl.

One caveat:
The UI path, i.e. for any generic auth provider or the OIDC auth provider in `fragment` mode will not send the expiration info or the refresh token. The reason for this is that they do not have this information nor can access it (since the cookie which holds the refresh token is `HttpOnly`, the auth saga may not read its contents and pass the refresh token to the roxctl server).

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

- added unit tests.
- manual tests:
1. Create an auth provider of your choice within Central (i.e. OIDC with fragment or query mode)
2. Run roxctl locally:
```bash
export ROX_ENDPOINT=<central hostname:port>
roxctl central login
```
3. You should see an URL being printed within the terminal. Additionally, the browser should open, and you should be redirected to a screen where you can select the auth provider with which you want to authorize roxctl
![Screenshot 2023-04-19 at 07 29 51](https://user-images.githubusercontent.com/89904305/232975818-a90329d1-b4ce-469b-9292-9c069282fbfa.png)

4. After you logged in with the auth provider, the browser should mention that authentication was successful, and you may now close the browser. Within the terminal, roxctl should print your token information and mention that it will be stored locally:
```bash
Please complete the authorization flow in the browser with an auth provider of your choice.
If no browser window opens, please click on the following URL:
        http://127.0.0.1:57767/login

INFO:	Received the following after the authorization flow from Central:
INFO:	Access token: <redacted
INFO:	Access token expiration: 2023-04-19 13:58:43 +0000 UTC
INFO:	Refresh token: <redacted>
INFO:	Storing these values under $HOME/.roxctl/login...
```

Screen record of the changes (with redacted token output):
https://user-images.githubusercontent.com/89904305/235177461-2e21f53e-60b1-4a27-a0e7-7854d8a2e693.mov
